### DISCUSSION TITLE
Error joining static worker

### DISCUSSION THREAD

**Discussion by cristianrat** [kubermatic/kubeone#2549] [Q&A]
So in short, I have a cluster on hetzner which works fine.
What I want is to join a metal box (ie: dedicated server) to this cluster.
What I tried:
* vswitch (ie: connect the subnet from dedicated to the cloud subnet) . Servers can see each other ping etc. Can't install - server never becomes healthy and remains "uninitialized"
* start a new cluster with a new cloud instance (ubuntu 20) and join this machine. Only public IPs, no private IP and no firewall 
It's stuck in `Waiting for nodes to initialize by CCM...   `
Looking at the logs, seems to be an issue with TLS, but unsure what to do about it
`http: TLS handshake error from 123.123.123.123:21909: no serving certificate available for the kubelet`

**Comment by kron4eg** [comment]
This happens because hetzner cloud cluster is working under the hetzner cloud controller manager (CCM), and this CCM obviously don't see your baremetal machine in the hetzner cloud API.

Only in theory we can configure kubelet on baremetal to avoid being uninitialized, but that will result in a mixed cluster and I'm very I'm far from certain that such cluster will work as intended and would recommend to avoid doing this.

**Reply by cristianrat** [reply]
Ah, I see
So really, what I need to do is configure my cluster as cloudProvider: none
That does mean losing out on volumes and other stuff (and also auto scaling)
Very annoying, this isn't specified in the docs, either - the docs show you how you can add static workers, so would be really nice to specify that one shouldn't mix-and-match

@kron4eg Would it be possible, then, for me to provision 3 cloud nodes, to be my control plane, and add what ever workers I want? (regardless of the cloud I would use), as long as I keep the cloudProvider to none?

**Reply by kron4eg** [reply]
@cristianrat yes, but I still wouldn't recommend it. Splitting a single cluster across the internet will work very bad.

**Reply by cristianrat** [reply]
@kron4eg wouldn't be split very far. same datacenter, just control nodes as cloud machines and workers as dedicated boxes

**Reply by kron4eg** [reply]
@cristianrat I've made a small patch (against current tip of the `main` branch). It will disable marking barematal machine as "uninitialized". TBH I don't know what will happen with such mixed cluster, but if you can and wish to try please do and please let us know how it will end up :)

```diff
$ git diff 
diff --git pkg/templates/kubeadm/v1beta3/kubeadm.go pkg/templates/kubeadm/v1beta3/kubeadm.go
index 09b987e8..faa28e93 100644
--- pkg/templates/kubeadm/v1beta3/kubeadm.go
+++ pkg/templates/kubeadm/v1beta3/kubeadm.go
@@ -443,11 +443,11 @@ func NewConfigWorker(s *state.State, host kubeoneapi.HostConfig) ([]runtime.Obje
                nodeRegistration.KubeletExtraArgs["cloud-config"] = renderedCloudConfig
        }
 
-       if cluster.CloudProvider.External {
-               if !s.ShouldEnableInTreeCloudProvider() {
-                       nodeRegistration.KubeletExtraArgs["cloud-provider"] = "external"
-               }
-       }
- 
        joinConfig.NodeRegistration = nodeRegistration
```

P.S.
How do you connect VMs and baremetal machines in a single flat network?

**Reply by cristianrat** [reply]
thanks for doing that @kron4eg . I'll try to give it a go, but not sure when (I presume that means I need to compile that specific version in order to get it to work?)

I connect the two via a vSwitch (ie: according to these instructions: https://docs.hetzner.com/cloud/networks/connect-dedi-vswitch/)

**Reply by kron4eg** [reply]
> I presume that means I need to compile that specific version in order to get it to work?

Yeah, you'd need to checkout the repo, apply the patch and compile a custom binary with `make build`.
